### PR TITLE
Sodium properties

### LINKED ISSUE #9717: Adding Sodium Fluid Properties to the FluidProperties Module
### Description of the enhancement or error report
Sodium Fluid Properties did not exist in the fluid properties module. Added SodiumProperty class deriving from FluidProperties base class. 

### Rationale for the enhancement or information for reproducing the error
Sodium fluid properties are needed for simulating sodium coolant channels. This used to be a free standing function in BISON. Now it goes along with the new input format for coolant channel model. 

### Identified impact
Now we can use sodium fluid properties.

### PR DESCRIPTION
**Pull request by ssatpat**
- Added SodiumProperty class derived from FluidProperties base class.
- Added regression test ( modules/fluid_properties/test/tests/sodium/ )
- Closes #9717

### CODE REVIEW

**Comment by brianmoose** [comment]
Not sure what is going on here, did you rebase on master by accident? Only your commits should be showing...also you will want to do a submodule update as you are using an old version of libmesh.

**Comment by andrsd** [comment]
@brianmoose I am helping him with this PR.

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+/*                                               */
+/*    (c) 2015 Battelle Energy Alliance, LLC     */
+/*            ALL RIGHTS RESERVED                */
+/*                                               */
+/*   Prepared by Battelle Energy Alliance, LLC   */
+/*     Under Contract No. DE-AC07-05ID14517      */
+/*     With the U. S. Department of Energy       */
+/*                                               */
+/*     See COPYRIGHT for full restrictions       */
+/*************************************************/
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: This needs MOOSE module header, not BISON's one.

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+#include "FluidProperties.h"
+#include "Moose.h"
+
+class SodiumProperty;
+
+template <>
+InputParameters validParams<SodiumProperty>();
+
+/// Properties of liquid sodium from ANL/RE-95/2 report "Thermodynamic and Transport Properties of
+/// Sodium Liquid and Vapor" from ANL Reactor Engineering Division.
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: Class doco uses `/** */` comments. Also no empty line between doco and `class`

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+{
+
+public:
+  SodiumProperty(const InputParameters & parameters);
+
+  /**
+   * Thermal conductivity as a function of temperature
+   * @param T temperature
+   */
+  Real thermal_conductivity(Real temperature) const;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: We use `k` for thermal conductivity in fluid properties module.

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   * Enthalpy of liquid Na (relative to solid Na at STP) in J/kg as a function of temperature
+   * @param T temperature
+   */
+  Real enthalpy(Real temperature) const;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: We use `h` for enthalpy in fluid properties module

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+   * Inverse solve for temperature [K] from enthalpy [J/kg]
+   * @param h enthalpy
+   */
+  Real temperature(Real enthalpy) const;
+
+  /**
+   * Density as a function of temperature
+   * @param T temperature
+   */
+  Real density(Real temperature) const;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: We use `rho` for density in fluid properties module

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+   * Enthalpy of liquid Na (relative to solid Na at STP) in J/kg as a function of temperature
+   * @param T temperature
+   */
+  Real enthalpy(Real temperature) const;
+
+  /**
+   * Heat capacity of liquid Na in J/kg-K as a function of temperature
+   * @param T temperature
+   */
+  Real heat_capacity(Real temperature) const;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: Is this heat capacity denoted by `C`? Or is this something else?

File: modules/fluid_properties/src/userobjects/SodiumProperty.C [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+/*                                               */
+/*    (c) 2015 Battelle Energy Alliance, LLC     */
+/*            ALL RIGHTS RESERVED                */
+/*                                               */
+/*   Prepared by Battelle Energy Alliance, LLC   */
+/*     Under Contract No. DE-AC07-05ID14517      */
+/*     With the U. S. Department of Energy       */
+/*                                               */
+/*     See COPYRIGHT for full restrictions       */
+/*************************************************/
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/SodiumProperty.C]: MOOSE module header

File: modules/fluid_properties/src/userobjects/SodiumProperty.C [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+  params.addClassDescription("Fluid properties for sodium");
+  return params;
+}
+
+SodiumProperty::SodiumProperty(const InputParameters & parameters) : FluidProperties(parameters) {}
+
+Real
+SodiumProperty::thermal_conductivity(Real T) const
+{
+  // From page 181. Valid from 371 to 1500 K.
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/SodiumProperty.C]: Move these report references + valid range comments into `.h`, they will be part of the generated doco and thus more useful.

File: modules/fluid_properties/src/userobjects/SodiumProperty.C [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  Real T = (H + 365.77e3) / 1.6582e3;
+
+  // Newton-Raphson for this equation: enthalpy(T) - H = 0 = residual. This is easy because
+  // dResidual/dT is just dH/dT, which is heat capacity.
+  for (unsigned iteration = 0; iteration < 10; ++iteration)
+  {
+    Real residual = enthalpy(T) - H;
+    T -= residual / heat_capacity(T);
+    if (std::abs(residual / H) < 1e-6)
+    {
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/SodiumProperty.C]: Do not need the block `{...}`.

File: modules/fluid_properties/src/userobjects/SodiumProperty.C [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+  // If we get here, enthalpy is probably out of bounds. However, due to the nature of the JFNK
+  // calculation, we probably just want to ignore the error and spit out a bogus T so that the
+  // solver keeps rolling.
+  return T;
+}
+
+Real
+SodiumProperty::density(Real T) const
+{
+  Real rho(0);
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/SodiumProperty.C]: Initialize variables using `=`, i.e. `Real rho = 0;`, etc.

File: modules/fluid_properties/src/userobjects/SodiumProperty.C [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  drho_dT = -(f + g * (0.5) / std::sqrt(1 - T / Tc)) / Tc;
+  return drho_dT;
+}
+
+Real
+SodiumProperty::drho_dh(Real h) const
+{
+  Real drho_dH(0);
+  Real T(0);
+  T = SodiumProperty::temperature(h);
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/SodiumProperty.C]: Why not `Real T = temperature(h);`?

File: modules/fluid_properties/test/tests/sodium/sodium.i [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+  nl_abs_tol = 1e-8
+
+  end_time = 2
+  dt = 0.5
+[]
+
+# Define output file(s)
+[Outputs]
+  exodus = true
+[]
```
Comment (andrsd) [inline modules/fluid_properties/test/tests/sodium/sodium.i]: This test cannot possibly run in fluid properties module. You are using app-specific syntax and classes from modules that are not dependencies of fluid properties. Go compile fluid properties module on your machine and execute `run_tests`. You should see the test fail.

Also, if you have unit tests for this new class (which would be preferred and easier actually), put them in `moose/unit`. That's where the classes from fluid properties module are unit-tested. You will see several there, so just follow the pattern.

Then you do not need this at all.

File: modules/fluid_properties/include/userobjects/SodiumProperty.h [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+/*     See COPYRIGHT for full restrictions       */
+/*************************************************/
+
+#ifndef SODIUMPROPERTY_H
+#define SODIUMPROPERTY_H
+
+#include "FluidProperties.h"
+#include "Moose.h"
+
+class SodiumProperty;
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/SodiumProperty.h]: We want to use plural, i.e. `SodiumProperties`, so we are consistent with the rest of the module.

**Review by andrsd — CHANGES_REQUESTED** [review]

File: modules/fluid_properties/include/userobjects/SodiumProperty.h (hunk 6 above)
Comment (ssatpat) [reply modules/fluid_properties/include/userobjects/SodiumProperty.h]: c is speed of sound as far as I remember. That's why heat_capacity is used here.

Comment (andrsd) [reply modules/fluid_properties/include/userobjects/SodiumProperty.h]: Right, I was referring to [this](https://en.wikipedia.org/wiki/Heat_capacity#Extensive_properties), they use capital `C` for it. I was not suggesting a rename to `C`. If you want to keep it spelled out, then we need to go with `camelCase`, not `c_style`, i.e. `heatCapacity`.  If this is `C_p` or `C_v`, then we would use `cp` or `cv`, respectively.

Comment (ssatpat) [reply modules/fluid_properties/include/userobjects/SodiumProperty.h]: Alright. I'll go with heatCapacity. I just find 'c' and 'C' confusing.

**Comment by andrsd** [comment]
You will need to get rid of the libmesh update too...

**Review by andrsd — DISMISSED** [review]

**Review by andrsd — APPROVED** [review]

**Pull request merged**